ambari-server/src/main/java/org/apache/ambari/server/checks/AbstractCheckDescriptor.java

```diff
@@ -468,4 +471,165 @@ public boolean isApplicable(PrereqCheckRequest request) throws AmbariException {
       return false;
     }
   }
+
+  /**
+   * Used to represent information about a service. This class is safe to use in
+   * sorted & unique collections.
+   */
+  static class ServiceDetail implements Comparable<ServiceDetail> {
+    @JsonProperty("service_name")
+    final String serviceName;
+
+    ServiceDetail(String serviceName) {
+      this.serviceName = serviceName;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int hashCode() {
+      return Objects.hash(serviceName);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean equals(Object obj) {
+      if (this == obj) {
```

Comment:
Seems every .equals() has this pattern, can we use EqualsBuilder(Object, Object) or similar to help here?

Reply:
I used `Objects.equals()` in these:

```
+      ServiceComponentDetail other = (ServiceComponentDetail) obj;
 +      return Objects.equals(serviceName, other.serviceName)
 +          && Objects.equals(componentName, other.componentName);
```

I figured the first few lines (comparing null, comparing referential equality, etc) wouldn't hurt. My IDE added them by default. Think I should remove them?

Reply:
Was just looking at EqualsBuilder.  The append(Object, Object) is doing `null`, `==`, and class equality checks, so I think it might be cleaner, but it's up to you and I won't hold up the review.

Reply:
`EqualsBuilder` and `Objects.equals()` both work the same way, with the exception being that `EqualsBuilder` uses method chaining. 

I'd rather use something built into the JDK at this point since i fear one day that we'll update Apache commons and `EqualsBuilder` will change or disappear.

The `this == obj` stuff would still need to exist in either implementation of `equals()`